banyand/series/series.go

```diff
@@ -22,29 +22,66 @@ import (
 	"context"
 
 	"github.com/apache/skywalking-banyandb/api/data"
+	v1 "github.com/apache/skywalking-banyandb/api/fbs/v1"
+	"github.com/apache/skywalking-banyandb/banyand/series/schema"
 	"github.com/apache/skywalking-banyandb/banyand/storage"
 	"github.com/apache/skywalking-banyandb/pkg/run"
 )
 
-type Trace interface {
+// TraceState represents the state of a trace link
+type TraceState int
+
+var (
+	TraceStateSuccess = 0
+	TraceStateError   = 1
+)
+
+//ScanOptions contain options
+//nolint
+type ScanOptions struct {
+	projection []string
+	state      TraceState
+	limit      uint32
+}
+
+//TraceRepo contains trace and entity data
+type TraceRepo interface {
+	//FetchTrace returns data.Trace by traceID
 	FetchTrace(traceID string) (data.Trace, error)
-	FetchEntity(chunkIDs []string, fields []string) ([]data.Entity, error)
-	ScanEntity(startTime, endTime uint64, fields []string) ([]data.Entity, error)
+	//FetchEntity returns data.Entity by ChunkID
+	FetchEntity(chunkIDs []string, opt ScanOptions) ([]data.Entity, error)
+	//ScanEntity returns data.Entity between a duration by ScanOptions
+	ScanEntity(startTime, endTime uint64, opt ScanOptions) ([]data.Entity, error)
 }
 
+//UniModel combines Trace, Metric and Log repositories into a union interface
 type UniModel interface {
-	Trace
+	TraceRepo
 }
 
+//SchemaRepo contains schema definition
+type SchemaRepo interface {
+	TraceSeries() schema.TraceSeries
+	IndexRule() schema.IndexRule
+	IndexRuleBinding() schema.IndexRuleBinding
+}
+
+//IndexFilter provides methods to find a specific index related objects
+type IndexFilter interface {
+	//RulesBySubject fetches IndexRule by Series defined in IndexRuleBinding
+	RulesBySubject(ctx context.Context, subject v1.Series) ([]v1.IndexRule, error)
```

Comment:
Add an API `FindRuleBySubjectAndField(s)` which returns a single IndexRule for the given field(s)

Reply:
A single field might belong to several index rules, for example, a field is indexed separately or belongs to a combinate index. so the returned rule should be a vector instead of a single item.

Reply:
I add a new parameter to filter a specific field(s), pls take a look